python/orca/src/bigdl/orca/data/shard.py

```diff
@@ -271,35 +270,24 @@ def repartition(self, num_partitions: int) -> "SparkXShards":
         if self._get_class_name() == 'pandas.core.frame.DataFrame':
             import pandas as pd
 
-            if num_partitions > self.rdd.getNumPartitions():
-                rdd = self.rdd \
-                    .flatMap(lambda df: df.apply(lambda row: (row[0], row.values.tolist()), axis=1)
-                             .values.tolist()) \
-                    .partitionBy(num_partitions)
-
-                schema = self.get_schema()
-
-                def merge_rows(iter):
-                    data = [value[1] for value in list(iter)]
-                    if data:
-                        df = pd.DataFrame(data=data, columns=schema['columns']) \
-                            .astype(schema['dtypes'])
-                        return [df]
-                    else:
-                        # no data in this partition
-                        return iter
-
-                repartitioned_shard = SparkXShards(rdd.mapPartitions(merge_rows))
-            else:
-                def combine_df(iter):
-                    dfs = list(iter)
-                    if len(dfs) > 0:
-                        return [pd.concat(dfs)]
-                    else:
-                        return iter
+            rdd = self.rdd \
+                .flatMap(lambda df: df.apply(lambda row: (row[0], row.values.tolist()), axis=1)
+                         .values.tolist()) \
+                .partitionBy(num_partitions)
 
-                rdd = self.rdd.coalesce(num_partitions)
```

Comment:
For Estimators that use Spark barrier mode, if input is spark xshards, we can't use coalesce for repartitions as coalesce can't work together with barrier mode.
```
py4j.protocol.Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.scheduler.BarrierJobUnsupportedRDDChainException: [SPARK-24820][SPARK-24821]: Barrier execution mode does not allow the following pattern of RDD chain within a barrier stage:
1. Ancestor RDDs that have different number of partitions from the resulting RDD (e.g. union()/coalesce()/first()/take()/PartitionPruningRDD). A workaround for first()/take() can be barrierRdd.collect().head (scala) or barrierRdd.collect()[0] (python).
2. An RDD that depends on multiple barrier RDDs (e.g. barrierRdd1.zip(barrierRdd2)).
```
@jason-dai @jenniew What's your opinion on this?
Even if coalesce can work, it will very likely to cause rather unbalanced partitions?

Reply:
use coalesce can reduce shuffling, but may cause unbalanced partitions.